src/Microsoft.DncEng.Configuration.Extensions/LocalDevTokenCredential.cs

```diff
@@ -0,0 +1,124 @@
+        }
+
+        private static string CacheDirectory { get; } =
+            Path.Join(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), ".auth-cache");
+
+        private static string CacheLockPath => Path.Join(CacheDirectory, ".cache.lock");
+
+        private static string CacheFilePath => Path.Join(CacheDirectory, "token.cache");
+
+        private static string DeviceCodeLockPath => Path.Join(CacheDirectory, ".auth.lock");
+
+        private static DateTimeOffset _lastReadTime;
+
+        private static async Task OnBeforeAccessAsync(TokenCacheNotificationArgs arg)
+        {
+            var ts = File.GetLastWriteTimeUtc(CacheFilePath);
+            if (File.Exists(CacheFilePath) && _lastReadTime < ts)
+            {
+                using (await SentinelFileLock.AcquireAsync(CacheLockPath, 100, TimeSpan.FromMilliseconds(600)))
+                {
+                    var protectedBytes = await File.ReadAllBytesAsync(CacheFilePath);
+                    var bytes = ProtectedData.Unprotect(protectedBytes, null, DataProtectionScope.CurrentUser);
+                    _lastReadTime = ts;
+                    arg.TokenCache.DeserializeMsalV3(bytes, false);
+                }
+            }
+        }
+
+        private static async Task OnAfterAccessAsync(TokenCacheNotificationArgs arg)
+        {
+            using (await SentinelFileLock.AcquireAsync(CacheLockPath, 100, TimeSpan.FromMilliseconds(600)))
+            {
+                _lastReadTime = default;
+                var bytes = arg.TokenCache.SerializeMsalV3();
+                var protectedBytes = ProtectedData.Protect(bytes, null, DataProtectionScope.CurrentUser);
+                await File.WriteAllBytesAsync(CacheFilePath, protectedBytes);
+            }
+        }
+
+        private IAccount _account;
+
+        public override async ValueTask<AccessToken> GetTokenAsync(TokenRequestContext requestContext, CancellationToken cancellationToken)
+        {
+            bool uiRequired = false;
+            if (_account != null)
+            {
+                try
+                {
+                    var result = await _app.AcquireTokenSilent(requestContext.Scopes, _account)
+                        .ExecuteAsync(cancellationToken);
+                    return new AccessToken(result.AccessToken, result.ExpiresOn);
+                }
+                catch (MsalUiRequiredException)
+                {
+                    // will be handled lower
+                    uiRequired = true;
+                }
+            }
+
+            // we give the lock a total wait time of 1 hour because the breakpoint to do device code authentication is inside the lock
+            using (await SentinelFileLock.AcquireAsync(DeviceCodeLockPath, 3600, TimeSpan.FromMilliseconds(1000)))
+            {
+                _account = Task.Run(async () => await _app.GetAccountsAsync(), cancellationToken).Result.FirstOrDefault();
```

Comment:
Why are we Task.Run a thing inside an async method?

Reply:
This was not in an async method initially. I will fix it.